### ISSUE TITLE
Segfault with fetch_add rmw kernel

### ISSUE THREAD

**Issue by seansiddens** [google/clspv#1170]
```OpenCL
__kernel void ticket_lock_test(__global atomic_uint* next_ticket, __global uint* now_serving) {
    uint my_ticket = atomic_fetch_add(next_ticket, 1);
    while (*now_serving != my_ticket) {}
}
```
causes a segfault on commit abeb0bb27cc83b2bcb5e19ddbffab3befbef4a5c when compiling with `--cl-std=CL2.0 --spv-version=1.3 --inline-entry-points`.

Godbolt link: https://godbolt.org/z/9WbnrvcY9

Noticed issue while implementing a ticket lock. The above snippet seems to be the smallest subset of that code which continued to segfault.

**Comment by rjodinchr** [comment]
This is the same bug as #708 

Infinite loops are not supported, if you add something to get out of the while loop in the example you submitted, it should compile correctly.

**Comment by seansiddens** [comment]
Why would changing `now_serving` to an atomic cause the kernel to compile correctly?

```OpenCL

__kernel void ticket_lock_test(__global atomic_uint* next_ticket, __global atomic_uint* now_serving) {
    uint my_ticket = atomic_fetch_add(next_ticket, 1);
    while (atomic_load(now_serving) != my_ticket) {}
}

```
Godbolt link: https://godbolt.org/z/6sYhszMbh

Both loops have a break condition relying on the variable being updated by some other thread.

**Comment by rjodinchr** [comment]
The first does not really have a break condition.
Because `now_serving` and `my_ticket` are not atomic, any compiler will compare them once, and after that the compiler will not reload those variables again. That's because according to the loop content the compiler knows/thinks that those 2 variables will not be updated from there.

When `now_serving` is turned into an atomic, it tells the compiler that this variable can be modified by something out of the reach of the compiler knowledge. Thus the compiler knows that it needs to evaluate the condition after executing the content of the loop. Here as there is nothing in the loop, it will just evaluate the condition over and over again, creating a break condition, not an infinite loop.

Basically what you need is to turn one of the variables of the condition as `volatlie`. By definition an atomic access is volatile. But volatile access is not atomic.

**Comment by rjodinchr** [comment]
closing the issue as a duplicate of #708
